\documentclass{article}
\usepackage[margin=1in]{geometry}
\usepackage{graphicx}
\pagestyle{empty}
\usepackage{floatrow}
\usepackage{subfig}
\usepackage{csvsimple}
\captionsetup[subfigure]{labelformat=simple,position=top,justification=justified,singlelinecheck=false}
\captionsetup[figure]{labelformat=simple,labelsep=space,labelfont=bf}

\title{snATAC and snRNA muscle manuscript figures}

\begin{document}

\floatsetup[figure]{style=plain,subcapbesideposition=top}

\maketitle
\pagestyle{empty}

  
\renewcommand{\thefigure}{\textbf{\arabic{figure}. }}
\setcounter{figure}{0}
	
	%\caption{ATAC-seq signal in bulk adipose, bulk islet, single-nucleus pancreatic beta cell, or our muscle cell types at the \textit{ITPR2} locus. Position of SNP rs7132434 is indicated by the long vertical red line. All ATAC-seq tracks are normalized to 1M reads and have the same y-axis range corresponding to 0 reads per million to 1 reads per million.}

\begin{figure}
\includegraphics[width=\textwidth]{figure-1}
	\caption{(A) Study design to determine the effect of FANS on snRNA-seq and snATAC-seq results. Muscle cartoon adapted from Scott et al. 2016. HSM1 refers to one specific skeletal muscle sample ('human skeletal muscle 1'). Bulk ATAC-seq was performed on HSM1 as well (two replicates, each separate nuclei isolations). (B) Fragment length distribution and (C) TSS enrichment for two snATAC-seq libraries that did not undergo FANS and two that did, as well as two bulk ATAC-seq replicates from the same sample ('Bulk'). (D) ATAC-seq signal at the \textit{ANK1} locus for FANS or non-FANS input snATAC-seq libraries, and the two bulk ATAC-seq libraries. All tracks are normalized to 1M reads and have the same y-axis range corresponding to 0 reads per million to 3 reads per million. Gene model (GENCODE v19 basic) displays protein coding genes only. (E) Correlation between FANS and non-FANS snRNA-seq libraries; each point represents one gene. (F) Study design to determine the effect of loading 20k vs 40k nuclei into the 10X platform, utilizing HSM1 as well as a second sample, HSM2 ('human skeletal muscle 2'). Bulk ATAC-seq was performed on HSM1 (same libraries as in (a)) and on HSM2 (two replicates, each separate nuclei isolations). (G) Fragment length distribution and (H) TSS enrichment for snATAC-seq libraries after loading 20k vs 40k nuclei, as well as for the four bulk ATAC-seq libraries (two each from the two muscle samples, 'HSM1 bulk' and 'HSM2 bulk'). (I) ATAC-seq signal at the \textit{ANK1} locus for the 20k and 40k libraries and the four bulk ATAC-seq libraries. All tracks are normalized as in (D). Gene model (GENCODE v19 basic) displays protein coding genes only. (J) Correlation between snRNA-seq libraries resulting from loading 20k vs 40k nuclei.}
\end{figure}

\begin{figure}
\includegraphics[width=\textwidth]{figure-2}
	\caption{(A) UMAP after clustering human snATAC-seq, human snRNA-seq, rat snATAC-seq, and human and rat dual modality (snATAC-seq+snRNA-seq) nuclei with Seurat. (B) UMAP facetted by species and modality. Dual modality nuclei were clustered using RNA and are displayed within the 'RNA' facets. (C) Gene expression (snRNA-seq, including dual modality nuclei RNA) or accessibility (snATAC-seq; gene promoter + gene body) of marker genes. Values are column-normalized. (D) ATAC-seq signal for human snATAC-seq (+ dual modality) nuclei in each cluster. All tracks are normalized to 1M reads. (E) Fraction of nuclei assigned to each cell type. (F) Logistic regression-based approach to score similarity between TSS-distal ATAC-seq peaks ($>$ 5 kb from TSS) and Roadmap Epigenomics enhancer states. For all TSS-distal ATAC-seq peaks across all cell types, we scored the accessibility of the peak (0/1) in each of the muscle cell types based on the presence or absence of a peak call. Then, for a given one of the 127 Roadmap Epigenomics cell types, we determined the maximum posterior probability of the enhancer states in the Roadmap Epigenomics chromHMM model within each peak. We then used logistic regression to model the relationship between the peak accessibility and the enhancer posteriors (running one model per muscle cell type per Roadmap Epigenomics cell type). Then, for each muscle cell type, the model coefficient was normalized to 1 by dividing by the maximum coefficient across all 127 Roadmap Epigenomics cell types, and this value was used as the enhancer similarity score for that muscle cell type and Roadmap Epigenomics cell type. (G) Similarity of snATAC-seq peak calls for each cell type and species to Roadmap Epigenomics chromHMM enhancer states based on the logistic regression procedure outlined in (F). The Roadmap Epigenomics cell type names have been adjusted for clarity and the sake of space. The full names and the identifiers from the Roadmap Epigenomics paper are: Psoas muscle (E100), Mesenchymal Stem Cell Derived Adipocyte Cultured Cells (E023), HUVEC Umbilical Vein Endothelial Primary Cells (E122), Stomach Smooth Muscle (E111), Primary monocytes from peripheral blood (E029), and Fetal Muscle Trunk (E089). (H) Nucleus counts per species for snATAC-seq data. Copyright disclosure: rat cartoon modified from https://commons.wikimedia.org/wiki/File:Vector\_diagram\_of\_laboratory\_mouse\_\%28black\_and\_white\%29.svg (by Gwilz [CC BY-SA 4.0 (https://creativecommons.org/licenses/by-sa/4.0)]).}
\end{figure}

\begin{figure}
	\includegraphics[width=0.4\textwidth]{label-color-legend}\\
\includegraphics[width=0.8\textwidth]{LDSC-UKB-across-all-cell-types.pdf}
	\caption{UK Biobank LDSC partitioned heritability results for traits for which at least one cell type showed significant heritability enrichment after Benjamini-Yekutieli correction (LDSC coefficient p $<$ 0.05) across all cell types and traits. Asterisks denote the significant cell type - trait combinations. One trait name has been shortened to preserve space: 'Diseases of veins, lymphatic vessels and lymph nodes, not elsewhere classified' has been shortened to 'Diseases of veins, lymphatic vessels and lymph nodes'}
\end{figure}

\begin{figure}
\includegraphics[width=0.85\textwidth]{figure-4}
	\caption{(A) LDSC partitioned heritability results for T2D (BMI-unadjusted) and Fasting insulin GWAS (BMI-adjusted), using human peak calls. For each cell type, one model was run adjusting for cell type-agnostic annotations from the LDSC baseline model and common open chromatin regions. Asterisks represents Bonferroni significance (p $<$ 0.05 after adjusting for 40 tests). (B) locuszoom plot for \textit{ITPR2} (T2D). (C) T2D credible set near the \textit{ITPR2} gene, consisting of 22 SNPs. Only SNP rs7132434 (highlighted in red) overlaps a peak call in any of the muscle cell types. (D) gkmexplain importance scores for the ref and alt allele (top two rows) and the difference between the ref and alt importance scores (third row); the G allele disrupts an AP1 motif (bottom row). (E). locuszoom plot for \textit{ARL15} locus (T2D). (F). T2D credible set SNPs near the \textit{ARL15} gene. The three SNPs represent the three-SNP credible set discussed in the text. One of these SNPs (rs702634; highlighted in red) overlaps a mesenchymal stem cell specific peak. (G). Projecting the SNP highlighted in (F), rs702634, into the rat genome (projected SNP position indicated by the red vertical line) shows the corresponding region has open chromatin in rat mesenchymal stem cells. (H). gkmexplain importance scores for the ref and alt alleles (top two rows), the difference between them (third row), and a MEF2 motif disrupted by the SNP. (I). Luciferase assay using construct containing either allele of SNP rs702634 in human adipose-derived mesenchymal stem cells. Each point represents one clone; the experiment was performed twice, once on two different days ('Replicate'). P-values computed using a two-sided unpaired t-test. Copyright disclosure: rat cartoon modified from https://commons.wikimedia.org/wiki/File:Vector\_diagram\_of\_laboratory\_mouse\_\%28black\_and\_white\%29.svg (by Gwilz [CC BY-SA 4.0 (https://creativecommons.org/licenses/by-sa/4.0)]).}
\end{figure}

% Supplemental figures
\renewcommand{\thefigure}{\textbf{S\arabic{figure}. }}
\setcounter{figure}{0}

\begin{figure}
\includegraphics[width=\textwidth]{fans-chromhmm-overlap}
	\caption{Chromatin state overlap for TSS-distal ($>$ 5kb from TSS) ATAC-seq peaks from the FANS and non-FANS snATAC-seq libraries.}
\end{figure}

\begin{figure}
\includegraphics[width=\textwidth]{fans-atac-correlation}
\caption{Correlation between FANS snATAC-seq, non-FANS snATAC-seq, and standard bulk ATAC-seq libraries. Each point represents one peak.}
\end{figure}

\begin{figure}
\includegraphics[width=\textwidth]{umis-vs-mitochondrial-fans-vs-no-fans}
\caption{QC thresholds for FANS and non-FANS snRNA-seq libraries. Dashed lines represent thresholds for minimum number of UMIs, maximum number of UMIs, and maximum fraction of mitochondrial UMIs.}
\end{figure}

\begin{figure}
\includegraphics[width=\textwidth]{loading-chromhmm-overlap}
	\caption{Chromatin state overlap for TSS-distal ($>$5 kb from TSS) ATAC-seq peaks from the 20k and 40k nucleus FANS snATAC-seq libraries.}
\end{figure}

\begin{figure}
\includegraphics[width=\textwidth]{loading-atac-correlation}
\caption{Correlation between 20k and 40k nucleus snATAC-seq libraries and standard bulk ATAC-seq libraries. Each point represents one peak.}
\end{figure}

\begin{figure}
	\subfloat[]{\includegraphics[width=0.6\textwidth]{hqaa-vs-tss-enrichment-20k-vs-40k}}\\
	\subfloat[]{\includegraphics[width=0.6\textwidth]{hqaa-vs-max-fraction-reads-from-single-autosome-20k-vs-40k}}
	\caption{QC thresholding for the 20k and 40k nuclei input snATAC-seq libraries. (a) Dashed lines represent thresholds for minimum number of reads, maximum number of reads, and minimum TSS enrichment. (b) Dashed lines represent thresholds for minimum number of reads, maximum number of reads, and the maximum fraction of reads derived from a single autosome (imposed to filter out nuclei showing abberant per-chromosome coverage).}
\end{figure} 

\begin{figure}
\includegraphics[width=0.6\textwidth]{umis-vs-mitochondrial-20k-vs-40k}
\caption{QC thresholds for the 20k and 40k nuclei input snRNA-seq libraries. Dashed lines represent thresholds for minimum number of UMIs, maximum number of UMIs, and maximum fraction of mitochondrial UMIs.}
\end{figure} 

\begin{figure}
	\subfloat[]{\includegraphics[width=0.8\textwidth]{hqaa-vs-tss-enrichment-used-downstream}}\\
	\subfloat[]{\includegraphics[width=0.8\textwidth]{hqaa-vs-max-fraction-reads-from-single-autosome-used-downstream}}
	\caption{QC thresholds for all snATAC-seq libraries used in cell type clustering and downstream analyses. (a) Dashed lines represent thresholds for minimum number of reads and minimum TSS enrichment. (b) Dashed lines represent thresholds for minimum number of reads and the maximum fraction of reads derived from a single autosome (imposed to filter out nuclei showing abberant per-chromosome coverage).}
\end{figure} 

\begin{figure}
\includegraphics[width=\textwidth]{umis-vs-mitochondrial-used-downstream}
\caption{QC thresholds for all snRNA-seq libraries used in cell type clustering and downstream analyses. Dashed lines represent thresholds for minimum number of UMIs, maximum number of UMIs, and maximum fraction of mitochondrial UMIs.}
\end{figure} 

\begin{figure}
	\includegraphics[width=\textwidth]{multiome-genome-comparisons}
	\caption{Number of ATAC fragments per 10X nuclear barcode mapping only to human or rat (left), number of UMIs per 10X nuclear barcode mapping only to human or rat (center), and fraction of species-unique ATAC fragments from human vs fraction of species-unique RNA UMIs from human in either nuclei called by DropletUtils' emptyDrops() function (Lun et al. 2019; top right) or nuclei passing imposed QC thresholds (middle and bottom right; middle shows all nuclei passing QC thresholds except the species specificity requirement; bottom shows all nuclei passing QC thresholds including successful assignment to a species). The greater deviation away from 0 or 1 and towards 0.5 along the RNA axis (y-axis) relative to the ATAC axis (x-axis) in the right-most panels is also observed in similar joint chromatin accessibility - gene expression data from another platform (Ma et al. 2020) and might be explained by the source of ambient RNA contamination being different than the source of ambient DNA contamination. The mixing rate noted in the upper-right panel is likely an upper bound estimate, as some nuclei marked as mixed may simply have higher ambient contamination; consistent with this hypothesis, many of the nuclei labeled here as 'mixed' still lean heavily towards one species (only 7\% of nuclei lie between 0.2 and 0.8 on the ATAC axis).}
\end{figure} 

\begin{figure}
	\includegraphics[width=0.6\textwidth]{dual-modality-qc}
	\caption{QC thresholds for multi-ome library}
\end{figure} 

\begin{figure}
\includegraphics[width=0.8\textwidth]{seurat-vs-liger-alluvial}
	\caption{Comparison between clustering with Seurat and clustering with LIGER. Cluster assignments are largely concordant (95.9\% of nuclei are assigned to the same cluster).}
\end{figure}

\begin{figure}
\includegraphics[width=0.5\textwidth]{fiber-types-per-species-MYH7-vs-MYH1_2_4}
	\caption{snATAC-seq fragment counts (gene promoter + gene body) and snRNA-seq read counts derived from the Type II muscle fiber myosin heavy chain genes (MYH1, MYH2, MYH4) or the Type I muscle fiber myosin heavy chain gene (MYH7) for human and rat nuclei. Each point represents a single nucleus. Type I muscle fibers/Type II muscle fibers headers represent the cluster to which each nucleus was assigned.}
\end{figure}

\begin{figure}
\includegraphics[width=\textwidth]{rubenstein-vs-our-fiber-type-lfcs}
	\caption{Log2(fold change) for Type II vs Type I muscle fiber gene expression, showing the genes with the largest fold changes between fiber types based on data from Rubenstein et al. (Rubenstein et al. Table S4). Rubenstein et al. performed RNA-seq on pooled type I and pooled type II muscle fibers, and determined the 20 genes with the largest fold change in type II relative to type I fibers, and the 20 genes with the largest fold change in the other direction, along with p-values for differential expression. The 34 genes (of those 40 genes) that were differentially expressed are shown here.}
\end{figure}

\begin{figure}
	\subfloat[]{\includegraphics[width=0.7\textwidth]{static-heatmap}}\\
	\subfloat[]{\includegraphics[width=0.3\textwidth]{{CEBPB_M4556_1.02}.png}}
	\subfloat[]{\includegraphics[width=0.3\textwidth]{{PAX7_M5710_1.02}.png}}
	\caption{Motif accessibility in muscle cell type clusters. (A) Mean motif chromVAR deviation z-score, normalized between 0 and 1 for each motif. Motifs relatively specific to a single cell type (see Methods) and with mean deviation z-score $>$ 2 in a cluster are shown; notable examples are labelled. An interactive, HTML version of this heatmap is available in Supplemental File S1. (B) and (C) Per nucleus deviation z-scores for two of the labelled motifs.}
\end{figure} 


\begin{figure}
	\includegraphics[width=0.4\textwidth]{label-color-legend}\\
	\includegraphics[width=0.7\textwidth]{UKB-LDSC-by-rn6}
	\caption{UK Biobank LDSC partitioned heritability results for traits for which at least one cell type showed significant heritability enrichment after Benjamini-Yekutieli correction (LDSC coefficient p $<$ 0.05) across all cell types and traits (rat peaks projected into human coordinates). Asterisks denote the significant cell type - trait combinations. One trait name has been shortened to preserve space: 'Diseases of veins, lymphatic vessels and lymph nodes, not elsewhere classified' has been shortened to 'Diseases of veins, lymphatic vessels and lymph nodes'}
\end{figure}

\begin{figure}
	\includegraphics[width=0.8\textwidth]{T2D-FIns-rn6}
	\caption{LDSC partitioned heritability results for T2D (BMI-unadjusted) and Fasting insulin GWAS (BMI-adjusted), using rat peak calls projected into human coordinates for the muscle cell types. Results are shown for pancreatic beta cell, adipose, and liver open chromatin regions as well. First, for each of the ten cell types, one model was run adjusting for cell type-agnostic annotations from the LDSC baseline model and common open chromatin regions (this is the joint model with open chromatin). Then, a single model containing those same annotations and all ten cell types was run (this is the joint model with open chromatin and all other cell types). Asterisk represents Bonferroni significance (p $<$ 0.05 after adjusting for two traits, ten cell types, and two models per cell type = 40 tests).} 
\end{figure}

\begin{figure}
	\includegraphics[width=\textwidth]{itpr2-human-big}
\includegraphics[width=\textwidth,trim=0em 29em 0em 0em,clip=true]{chrom_state_legend}
	\caption{ATAC-seq signal in bulk adipose, bulk islet, single-nucleus pancreatic beta cell, or our muscle cell types at the \textit{ITPR2} locus. Position of SNP rs7132434 is indicated by the long vertical red line. All ATAC-seq tracks are normalized to 1M reads and have the same y-axis range corresponding to 0 reads per million to 1 reads per million.}
\end{figure}

\begin{figure}
	\includegraphics[width=0.5\textwidth]{dsvm-vs-allelic-bias}
	\caption{Comparison between deltaSVM z-score and ATAC-seq allelic bias directionality, in type II muscle fiber nuclei. Positive deltaSVM scores indicate greater predicted accessibility for the alt. allele. SNPs shown are those that have absolute deltaSVM z-score of at least 2, and show ATAC-seq allelic bias (FDR $<$ 1\%, using binomial test w/ expected fraction ref. = 0.5; only SNPs with min. coverage = 15 and at least one ref and one alt allele observed were tested)}
\end{figure}

\begin{figure}
	\includegraphics[width=\textwidth]{12_26453283}
	\caption{DeltaSVM z-score for each cell type and credible SNP at the ITPR2 locus.}
\end{figure}

\begin{figure}
	\subfloat[]{\includegraphics[width=0.5\textwidth]{FOS}}\\
	\subfloat[]{\includegraphics[width=0.5\textwidth]{JUN}}
	\caption{Epression of two AP-1 subunits in each cluster (human).}
\end{figure} 


\begin{figure}
	\includegraphics[width=0.7\textwidth]{ITPR2_peak-connections}
	\caption{Presence (dark red) or absence (light red) of Hi-C connection between ATAC-seq peak containing SNP rs7132434 and gene TSS within 1 Mb of the SNP, across 21 tissues and cell types profiled in from Schmitt et al. 2016.}
\end{figure}

\begin{figure}
	\includegraphics[width=0.4\textwidth]{gene-expression-near-ITPR2-locus}
	\caption{Expression (log(CPM + 1)) of genes with TSS within 1 Mb of SNP rs7132434}
\end{figure}

\begin{figure}
	\includegraphics[width=0.4\textwidth]{ITPR2-cicero}
	\caption{CICERO co-accessibility scores between ATAC-seq peak containing SNP rs7132434 and gene TSS within 1 Mb of the SNP. Grey = missing data (e.g., no gene TSS peak or CICERO coaccessibility of 'NA').}
\end{figure}

\begin{figure}
	\includegraphics[width=0.3\textwidth]{ITPR2-signac}
	\caption{Z-score for peak - gene expression correlation (calculated using Signac software package LinkPeaks function), using signal of ATAC-seq peak containing SNP rs7132434 and genes within 1 Mb of the SNP. Z-score represents Spearman correlation of the peak - gene combination relative to correlation between the gene and control peaks. Grey = missing data (gene - peak pair not tested, e.g. because fewer than 5 nuclei showed a read count for both the gene and the peak).}
\end{figure}

\begin{figure}
\includegraphics[width=\textwidth]{arl15-human-big}
\includegraphics[width=\textwidth,trim=0em 29em 0em 0em,clip=true]{chrom_state_legend}
	\caption{ATAC-seq signal in bulk adipose, bulk islet, single-nucleus pancreatic beta cell, or our muscle cell types at the \textit{ARL15} locus. Position of SNP rs702634 is indicated by the long vertical red line. All ATAC-seq tracks are normalized to 1M reads and have the same y-axis range corresponding to 0 reads per million to 1 reads per million.}
\end{figure}

\begin{figure}
	\includegraphics[width=0.8\textwidth]{5_53271420}
	\caption{DeltaSVM z-score for each cell type and credible SNP at the ARL15 locus.}
\end{figure}


\begin{figure}
	\subfloat[]{\includegraphics[width=0.4\textwidth]{MEF2A}}
	\subfloat[]{\includegraphics[width=0.4\textwidth]{MEF2C}}\\
	\subfloat[]{\includegraphics[width=0.4\textwidth]{MEF2D}}
	\caption{MEF2A/C/D expression in each cluster (human). Due to an overlapping transcript annotation, MEF2B expression is not quantified.}
\end{figure} 


\begin{figure}
	\includegraphics[width=0.75\textwidth]{luciferase-all}
	\caption{Luciferase assay using construct containing either allele of SNP rs702634, in human adipose-derived mesenchymal stem cells, pre-adipocytes (24 hours of mesenchymal stem cell adipogenic differentiation), and adipocytes (11 days of adipogenic mesenchymal stem cell differentiation). Each point represents one clone; p-values computed using a two-sided unpaired t-test.}
\end{figure}

\begin{figure}
	\includegraphics[width=0.8\textwidth]{ARL15_peak-connections}
	\caption{Presence (dark red) or absence (light red) of HiC connection between ATAC-seq peak containing SNP rs702634 and gene TSS within 1 Mb of the SNP, across 21 tissues and cell types profiled in from Schmitt et al. 2016.}
\end{figure}

\begin{figure}
	\includegraphics[width=0.7\textwidth]{ARL15-cicero}
	\caption{CICERO co-accessibility scores between ATAC-seq peak containing SNP rs702634 and gene TSS within 1 Mb of the SNP. Grey = missing data (e.g., no gene TSS peak or CICERO coaccessibility of 'NA').}
\end{figure}

\begin{figure}
	\includegraphics[width=0.4\textwidth]{gene-expression-near-ARL15-locus}
	\caption{Expression (log(CPM + 1)) of genes with TSS within 1 Mb of SNP rs702634}
\end{figure}

\begin{figure}
	\includegraphics[width=0.3\textwidth]{ARL15-signac}
	\caption{Z-score for peak - gene expression correlation (calculated using Signac software package LinkPeaks function), using signal of ATAC-seq peak containing SNP rs702634 and genes within 1 Mb of the SNP. Z-score represents Spearman correlation of the peak - gene combination relative to correlation between the gene and control peaks. Grey = missing data (gene - peak pair not tested, e.g. because fewer than 5 nuclei showed a read count for both the gene and the peak).}
\end{figure}


\begin{figure}
	\subfloat[]{\includegraphics[width=0.7\textwidth]{63_20_rna-hg19}}\\
	\subfloat[]{\includegraphics[width=0.7\textwidth]{63_40_rna-hg19}}
	\caption{Barcode rank plots for snRNA-seq libraries with (A) 20k or (B) 40k nuclei loaded. Horizontal red line represents 1000 UMIs (min UMI threshold for a nucleus to be called). Vertical red line represents the rank corresponding to the number of nuclei ostensibly loaded}
\end{figure}



\end{document}
